:running: Design for Simplified Scaffolding

DirectXMan12 opened the pull request:
This introduces a design for a simplified scaffolding that addresses
both persistent pedagogical issues and dependency injection issues with
our current scaffolding.

DirectXMan12 commented:
see also #603.  Holding off on a full-on prototype for a bit, but this looks quite a bit similar to how I structure the controllers when I'm developing or prototyping with CR.

DirectXMan12 commented:
cc @droot @mengqiy

mengqiy commented on designs/simplified-scaffolding.md:
```diff
@@ -0,0 +1,512 @@
+│   ├── mykind_controller_test.go
+│   └── controllers_suite_test.go
+├── api
+│   └── v1
```
It sounds like this is the proposed structure if we have more than one versions.
```
├── api
│   ├── v1
│   │   └── foo_types.go
│   │   └── groupversion_info.go
│   └── v2
│       └── foo_types.go
│       └── groupversion_info.go
```
It sounds like the info about group will be duplicated across versions.
Given we assume only one group in this structure, does it make sense to have structure like
```
├── api
│   ├── group_info.go
│   ├── v1
│   │   └── foo_types.go
│   │   └── version_info.go
│   └── v2
│       └── foo_types.go
│       └── version_info.go
```

DirectXMan12 replied:
good call

mengqiy commented on designs/simplified-scaffolding.md:
```diff
@@ -0,0 +1,512 @@
+3. Tell users to move things, and scaffold out with the new structure.
+   This is fairly messy for the user.
+
+Since growing to multiple API groups seems to be fairly uncommon, it's
```
> fairly uncommon

IIUC you mentioned that you have some data to back this when we chat offline. If you put the number here it's more convincing to make the decision :)

DirectXMan12 replied:
swear I had this somewhere, but can't seem to find the data now.  I'll see what I can put together

DirectXMan12 replied:
hmm... reproducing what I had before is proving difficult (the query I was using before is now returning wildly truncated results), but with a random sampling of smaller projects on github using kubebuilder (presence of manager.yaml) and larger projects that I know of that make use of controller runtime or just CRDs in general (e.g. kubeflow, knative, cluster-api, cert-manager, app-crd, prometheus operator, etcd operator) most seem to follow the pattern of 1 API group per repo.  The main exception was a couple of the knative repositories (but not all of them) and openshift, which puts all of its types in a single repo anyway, so isn't relevant to us.

This makes some amount of sense -- most operators or controllers logically operate on a single set of resources (the CRD defining the operator, a collection of related resources that work together) which would generally be grouped into a single API group.

droot replied:
hmm...another crazy idea :)

What if we removed the `api` and introduced the *group* hierarchy.

```shell
./test/project
  - groupa/v1/....
  - groupb/v2/...
```

droot commented on designs/simplified-scaffolding.md:
```diff
@@ -0,0 +1,512 @@
+(`main.go` is the defacto standard entrypoint for many go programs), and
+simplifies the levels of hierarchy.  Furthermore, since `main.go` actually
+instantiates an instance of the reconciler, users are able to add custom
+logic having to do with flags.
```
+1

droot commented on designs/simplified-scaffolding.md:
```diff
@@ -0,0 +1,512 @@
+   the current structure.
+
+Option 1 should be fairly simple, since the logic is already needed for
+registering types to the scheme.
```
In case where user has removed that particular code comment (or we are unable to wire second controller for some reason), we can always fallback to option 2 where we just emit the snippet which user can copy-paste in their main.go.

DirectXMan12 commented:
> What if we removed the api and introduced the group hierarchy.

hmm... in this case, would controllers go under this hierarchy as well?  I'm not certain this would be clear, depending on the group name -- e.g. with a group name of `apps` or `dependencies`, it's not clear that it's an API group and not something else.

droot commented:
>hmm... in this case, would controllers go under this hierarchy as well? I'm not certain this would be clear, depending on the group name -- e.g. with a group name of apps or dependencies, it's not clear that it's an API group and not something else.

Yeah, removing `api` and introducing `groups` directly can get confusing.
